api/types/resetpasswordtoken.go

```diff
@@ -158,3 +175,32 @@ func (u ResetPasswordTokenV3) CheckAndSetDefaults() error {
 func (u *ResetPasswordTokenV3) String() string {
 	return fmt.Sprintf("ResetPasswordTokenV3(tokenID=%v, user=%v, expires at %v)", u.GetName(), u.Spec.User, u.Expiry())
 }
+
+// IncrementAuthAttempt increases the attempt counter by 1.
+func (u *ResetPasswordTokenV3) IncrementAuthAttempt() {
+	u.AuthAttempts += 1
+}
+
+// GetAuthAttempts returns number of failed auth attempts.
+func (u *ResetPasswordTokenV3) GetAuthAttempts() int32 {
+	return u.AuthAttempts
+}
+
+// CheckSubKindType checks if given token type matches with the subkind.
+func (u *ResetPasswordTokenV3) CheckSubKindType(tokenType string) error {
```

Comment:
This validation appears too ad hoc. The `tokenType` should in some way be part of `ResetPasswordToken` itself which would make these checks a matter of internal consistency that can be verified immediately upon creation of a token object. E.g., would it make sense to store the request used to create a token within the token object itself?

Or, the function could be refactored into a ```func (u *ResetPasswordTokenV3) GetType() (string, error)``` that is then checked by the caller instead of the caller passing a string in.

Reply:
I removed this check. The token spec has now a field called `RecoveryType` that stores this type of data.